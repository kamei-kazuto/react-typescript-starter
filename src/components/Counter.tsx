import * as React from 'react'
import { counterStore } from '@/store'
export interface Props {
  count: number
  increment?: () => void
  decrement?: () => void
  reset?: () => void
}

export const Component: React.SFC<Props> = ({
  count,
  increment,
  decrement,
  reset
}) => {
  const hancleClick = (action: () => void) => (
    e: React.MouseEvent<HTMLButtonElement>
  ) => action()
  return (
    <div>
      <h3>COUNTER</h3>
      <div>{count}</div>
      <button onClick={hancleClick(increment)}>+1</button>
      <button onClick={hancleClick(decrement)}>-1</button>
      <button onClick={hancleClick(reset)}>Reset</button>
    </div>
  )
}

export const Counter: React.SFC<{}> = () => {
  const {
    state: { count },
    increment,
    decrement,
    reset
  } = counterStore()

  return (
    <Component
      count={count}
      increment={increment}
      decrement={decrement}
      reset={reset}
    />
  )
}
