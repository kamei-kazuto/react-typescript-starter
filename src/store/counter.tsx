import * as React from 'react'

interface State {
  count: number
}
interface Action {
  type: ActionTypes
  param: any
}
enum ActionTypes {
  INCREMENT = 'increment',
  DECREMENT = 'decrement',
  RESET = 'reset'
}
export interface Store {
  state: State
  increment: () => void
  decrement: () => void
  reset: () => void
}
const initialState: State = {
  count: 0
}

const CounterContext = React.createContext({
  state: initialState,
  dispatch: undefined
})

const reducer = (
  state: State = initialState,
  { type, param }: Action
): State => {
  switch (type) {
    case ActionTypes.INCREMENT:
      return {
        ...state,
        count: state.count + 1
      }
    case ActionTypes.DECREMENT:
      return {
        ...state,
        count: state.count - 1
      }
    case ActionTypes.RESET:
      return {
        ...state,
        count: 0
      }
    default:
      return state
  }
}

export const CounterProvider: React.SFC<{}> = ({ children }) => {
  const [state, dispatch] = React.useReducer(reducer, initialState)
  const { Provider } = CounterContext
  return (
    <Provider
      value={{
        state,
        dispatch
      }}
    >
      {children}
    </Provider>
  )
}

export const counterStore = (): Store => {
  const { state, dispatch } = React.useContext(CounterContext)
  const increment = () =>
    dispatch({
      type: ActionTypes.INCREMENT
    })
  const decrement = () =>
    dispatch({
      type: ActionTypes.DECREMENT
    })
  const reset = () =>
    dispatch({
      type: ActionTypes.RESET
    })
  return {
    state,
    increment,
    decrement,
    reset
  }
}
