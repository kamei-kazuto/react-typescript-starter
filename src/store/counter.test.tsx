import * as React from 'react'
import { shallow, mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import { CounterProvider, counterStore } from './counter'
import { Counter as Container } from '@/components/Counter'

describe('CounterProvider', () => {
  const render = mount(
    <CounterProvider>
      <Container />
    </CounterProvider>
  )
  it('snapshot', () => {
    expect(toJson(render)).toMatchSnapshot()
  })
  it('Component props test', () => {
    const Component = render.children().children()
    expect(Component.prop('count')).toBe(0)
    expect(Component.prop('increment')).toBeDefined()
    expect(Component.prop('decrement')).toBeDefined()
    expect(Component.prop('reset')).toBeDefined()
  })
})
