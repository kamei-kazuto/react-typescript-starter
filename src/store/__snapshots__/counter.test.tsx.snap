// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CounterProvider snapshot 1`] = `
<Component>
  <Component>
    <Component
      count={0}
      decrement={[Function]}
      increment={[Function]}
      reset={[Function]}
    >
      <div>
        <h3>
          COUNTER
        </h3>
        <div>
          0
        </div>
        <button
          onClick={[Function]}
        >
          +1
        </button>
        <button
          onClick={[Function]}
        >
          -1
        </button>
        <button
          onClick={[Function]}
        >
          Reset
        </button>
      </div>
    </Component>
  </Component>
</Component>
`;
